File: 2021/day6/lanternFish2.py


def readFile(filename):
    file = open(filename, 'r')
    lines = file.readlines()
    return lines[0]

if __name__ == '__main__':
    initialState = [int(x) for x in readFile("input").strip().split(',')]
    fishStates = [0 for _ in range(9)]
    
    for i in range(9):
        fishStates[i] = initialState.count(i)


    print(f"{fishStates}")
    for day in range(256):
        print(f"{day} {fishStates}")
        updateFishStates = [0 for _ in range(9)]
        for i in range(9):
            if i == 0:
                updateFishStates[6] += fishStates[i]
                updateFishStates[8] += fishStates[i]
            else:
                updateFishStates[i-1] +=fishStates[i]
        fishStates = updateFishStates.copy()
        print(f"!{day} {fishStates}")
        print(f"?{day} {updateFishStates}")

    print(f"{fishStates}")
    print(f"{sum(fishStates)}")


            

File: 2021/day6/lanternFish2.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <vector>

std::stringstream readFile(char * filename)
{
    std::ifstream inFile;
    std::string line;

    inFile.open(filename);
    std::getline(inFile, line);
    printf("%s", line.c_str());
    std::stringstream iss(line);

    return iss;
}

int main()
{
    std::vector<int> initialStateList;
    std::stringstream initialState = readFile((char *) "input");

    for (int k; initialState >> k;) {
        initialStateList.push_back(k);
        if (initialState.peek() == ',')
            initialState.ignore();
    }
    for (std::size_t p = 0; p < initialStateList.size(); p++)
        std::cout << initialStateList[p] << std::endl;   

    for (int day = 0; day < 256; day++)
    {
        //#listLength += spawnCount;
        printf("day: %d | list size: %ld\n",day,initialStateList.size());
        for (unsigned long long i = 0; i < initialStateList.size(); i++)
	{
	    if (initialStateList[i] == 0)
            {
                initialStateList[i] = 6;
		initialStateList.push_back(8);
		continue;
            }
    	    initialStateList[i]--;
	}
    }
    printf("%ld",initialStateList.size());
    return 1; 
}
